### PR TITLE
feat: adds iterable support to pagination

### PR DESCRIPTION
**Pull request by tdstein**
I'm working on a change to make Paginator a mixin which adds iterable support. Currently, the paginator is a property of Users, but the end state will be `class Users(Paginator[User])`. This will add support to Users for `iter`, `next`, and `len` builtins.

### CODE REVIEW

**Comment by tdstein** [comment]
Once this is stable, it might be a good time to add the rest of the paginated API endpoints.

**Comment by tdstein** [comment]
It looks like there are 6 total. Searching for page_number here: https://docs.posit.co/connect/api/swagger.json

File: src/posit/connect/paginator.py [hunk 1]
```diff
@@ -1,49 +1,111 @@
-from typing import List
+from collections.abc import Iterable
+from typing import Iterator, List, Tuple
 
+import requests
 
-# The maximum page size supported by the API.
-_MAX_PAGE_SIZE = 500
```
Comment (nealrichardson) [inline src/posit/connect/paginator.py:5]: Is this not constant across all paginated APIs?

File: src/posit/connect/paginator.py [hunk 2]
```diff
@@ -1,49 +1,111 @@
...
+    def __len__(self) -> int:
+        """
+        Returns the total number of items in the paginator.
+
+        Returns:
+            int: The total number of items in the paginator.
+        """
+        # Define query parameters for pagination.
+        params = {"page_number": 1, "page_size": 1}
+        # Send a GET request to the endpoint with the specified parameters.
```
Comment (nealrichardson) [inline src/posit/connect/paginator.py:81]: I think you can delete these prompt-comments, they don't add any useful to the code

File: src/posit/connect/paginator.py [hunk 3]
```diff
@@ -1,49 +1,111 @@
...
+        return self
+
+    def __len__(self) -> int:
+        """
+        Returns the total number of items in the paginator.
+
+        Returns:
+            int: The total number of items in the paginator.
+        """
+        # Define query parameters for pagination.
```
Comment (nealrichardson) [inline src/posit/connect/paginator.py:79]:
```suggestion
        # Connect's paginated APIs include a "total" count in the response.
        # So we only need to request page_size = 1 and inspect the total.
```

File: src/posit/connect/paginator.py [hunk 4]
```diff
@@ -1,49 +1,111 @@
...
+        params = {"page_number": 1, "page_size": 1}
```
Comment (nealrichardson) [inline src/posit/connect/paginator.py:80]: Does `page_size=0` work?

File: src/posit/connect/users.py [hunk 5]
```diff
@@ -25,32 +25,73 @@ class User(TypedDict, total=False):
...
+        *,
+        page_size: int = _MAX_PAGE_SIZE,
+    ) -> Optional[User]:
+        # You may be tempted to to use self.find(filter, page_size) and return the first element.
+        # This is less efficient as it will fetch all the users and then filter them.
+        # Instead, we use the paginator directly to fetch the first user that matches the filter.
+        # Since the paginator fetches the users in pages, it will stop fetching subsequent pages once a match is found.
+        url = urls.append_path(self.config.url, "v1/users")
+        paginator = Paginator(self.session, url, page_size=page_size)
+        users = (User(**user) for user in paginator)
```
Comment (nealrichardson) [inline src/posit/connect/users.py:90]: Reading this, it's not obvious to me that each of these comprehensions isn't actually iterating over all of the results in the paginator. (There at least used to be versions of Python where that wouldn't be true, unless I'm misremembering.)

So for me, that merits (a) an inline comment acknowledging how it works, and (b) a `responses`-based test that would error if you went beyond a certain page of results. The current test doesn't ensure that, IIRC.

**Review by nealrichardson — COMMENTED** [review]
A few notes here.

This is cool, but I'm wondering how it helps. To my eyes, it's less immediately obvious how the pagination works in practice with this implementation--the previous one is much less clever. So what do we get for this extra complexity, and is it worth it? It may very well be worth it, I just thought it worth asking.

File: src/posit/connect/paginator.py (hunk 2 above)
Comment (tdstein) [reply src/posit/connect/paginator.py:81]: Yeah, it's interesting. I've done this a few times in different places and have seen mixed results. Some have found them very useful, and others haven't. You can get the same information from `/explain` in GitHub Copilot, but that's a new workflow concept for most folks.

File: src/posit/connect/paginator.py (hunk 4 above)
Comment (tdstein) [reply src/posit/connect/paginator.py:80]: Unfortunately, no, `posit.connect.errors.ClientError: The parameter page_size=0 is less than the allowed minimum 1 (Error Code: 25, HTTP Status: 400 Bad Request)`.

File: src/posit/connect/users.py (hunk 5 above)
Comment (tdstein) [reply src/posit/connect/users.py]: I added some comments to try and explain the behavior.

**Comment by tdstein** [comment]
> A few notes here.
> 
> This is cool, but I'm wondering how it helps. To my eyes, it's less immediately obvious how the pagination works in practice with this implementation--the previous one is much less clever. So what do we get for this extra complexity, and is it worth it? It may very well be worth it, I just thought it worth asking.

Thanks! I want to avoid unnecessary complexity, so happy to iron this out more.

Could you expand on which parts you find complex?

I have a few goals in mind.

1. Allow idiomatic comprehension on resource collections: `(user for user in client.users)`.
1. Provide mixin support for other resource types.
1. Make the comprehension memory efficient by waiting to fetch subsequent pages until they are needed.

As we've been discussing. The memory efficiency isn't premature since endpoint support is limited. But I'm hoping that this can encourage the migration of other endpoints which manage large collections.

File: src/posit/connect/paginator.py [hunk 6]
```diff
@@ -1,49 +1,103 @@
...
+        return self
+
+    def __len__(self) -> int:
+        """
+        Returns the total number of items in the paginator.
+
+        Returns:
+            int: The total number of items in the paginator.
+        """
+        # Connect's paginated APIs include a "total" count in the response.
```
[the remark this answers is not in the document]
Comment (tdstein) [reply src/posit/connect/paginator.py:73]: Thanks for pointing this out. I wasn't aware of this API yet. Yes, we can create two different `Paginatior` implementations. Add https://github.com/posit-dev/posit-sdk-py/issues/75

**Comment by tdstein** [comment]
> * I don't think that it's safe to lazily take pages to feed the iterator. The content API at least is sorted by "recently updated", right? So if I iterate through that, make updates to some or all of the items in it, and fetch pages lazily, the contents of e.g. "page_number=2" may be changing as a result of my action. Users seem to be sorted alphabetically. So if paginated results are sorted on mutable properties, we may get surprising results unless we fetch all before allowing actions that could mutate the server. Even if sorting on immutable properties, the ability to delete entries will make page contents unstable.

I see this as a server-side implementation issue rather than a client-side one. Without consistent ordering and support for pagination tokens, the underlying dataset can change regardless of the time between page fetches.

**Comment by nealrichardson** [comment]
> > * I don't think that it's safe to lazily take pages to feed the iterator. The content API at least is sorted by "recently updated", right? So if I iterate through that, make updates to some or all of the items in it, and fetch pages lazily, the contents of e.g. "page_number=2" may be changing as a result of my action. Users seem to be sorted alphabetically. So if paginated results are sorted on mutable properties, we may get surprising results unless we fetch all before allowing actions that could mutate the server. Even if sorting on immutable properties, the ability to delete entries will make page contents unstable.
> 
> I see this as a server-side implementation issue rather than a client-side one. Without consistent ordering and support for pagination tokens, the underlying dataset can change regardless of the time between page fetches.

Sure, it's a server-side issue, but supposing we change/fix it there, the SDK still has to support the current API. We can only hope to minimize its impact by fetching all at the ~same time (AFAICT).

**Comment by tdstein** [comment]
I'm going to abandon this idea for the time being. We can revisit if we decide to make resources iterable in the future.

**Pull request closed without merging**